### PR TITLE
Emit metrics in metadata consumer subscriber service

### PR DESCRIPTION
**Pull request by snehalakshmisha**
Testing:
`POST` on `metrics/query` with payload
```
{ "query": { "tags": { "ns" : "default" }, "metrics": [ "system.metadata.consumer.calls.attempted"], "timeRange": { "aggregate": true }}}
```
gave the response below
```
{
  "query": {
    "startTime": 0,
    "endTime": 1678930364,
    "series": [
      {
        "metricName": "system.metadata.consumer.calls.attempted",
        "grouping": {},
        "data": [
          {
            "time": 0,
            "value": 4
          }
        ]
      }
    ],
    "resolution": "2147483647s"
  }
}
```

Jira: https://cdap.atlassian.net/browse/CDAP-20451

### CODE REVIEW

File: cdap-metadata-spi/src/main/java/io/cdap/cdap/spi/metadata/MetadataConsumerContext.java [hunk 1]
```diff
@@ -34,4 +34,9 @@ public interface MetadataConsumerContext {
    * @return the MetadataConsumer properties for the program run
    */
   Map<String, String> getProperties();
+
+  /**
+   * Returns a {@link MetadataConsumerMetrics} object configured based on {@code context}.
+   */
+  MetadataConsumerMetrics getMetrics(Map<String, String> context);
```
Comment (greeshmaswaminathan) [inline cdap-metadata-spi/src/main/java/io/cdap/cdap/spi/metadata/MetadataConsumerContext.java:41]: Can MetricsContext be used directly here? Is there a problem with introducing `cdap-watchdog-api` dependency to the SPI?

Comment (snehalakshmisha) [reply cdap-metadata-spi/src/main/java/io/cdap/cdap/spi/metadata/MetadataConsumerContext.java:41]: I used this commit for reference - https://github.com/cdapio/cdap/commit/c90bcfae320b009b61eac368b607afa20b5ab419.
This is for a similar case, for dataproc runtime-spi. They have their own implementation of metrics here

File: cdap-common/src/main/java/io/cdap/cdap/common/conf/Constants.java
Comment (rmstar) [inline cdap-common/src/main/java/io/cdap/cdap/common/conf/Constants.java:1]: What changed in this file? I am just seeing checkstyle fixes

File: cdap-metadata-spi/src/main/java/io/cdap/cdap/spi/metadata/MetadataConsumerMetrics.java [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+ */
+public interface MetadataConsumerMetrics {
+  
+  /**
+   * Increases the value of the specific metric by delta.
+   *
+   * @param metricName Name of the counter. Use alphanumeric characters in metric names.
+   * @param delta The value to increase by.
+   */
+  void count(String metricName, int delta);
```
Comment (greeshmaswaminathan) [inline cdap-metadata-spi/src/main/java/io/cdap/cdap/spi/metadata/MetadataConsumerMetrics.java]: Can we make this consistent with `io.cdap.cdap.api.metrics.MetricsCollector` increment() and use String, long as arguments ?

File: cdap-common/src/main/java/io/cdap/cdap/common/conf/Constants.java
Comment (snehalakshmisha) [reply cdap-common/src/main/java/io/cdap/cdap/common/conf/Constants.java:1]: Just this line - https://github.com/cdapio/cdap/pull/14987/files#diff-f27d32c0ca8fad3daa2a7437644e084d96983189ebff676ab5f2ce1b496e2d29R989 :)

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/metadata/MetadataConsumerSubscriberService.java [hunk 3]
```diff
@@ -265,35 +265,34 @@ public void processMessage(MetadataMessage message, StructuredTableContext conte
...
         return;
       }
       this.consumers.forEach((key, consumer) -> {
         try {
           DefaultMetadataConsumerContext metadataConsumerContext =
-              new DefaultMetadataConsumerContext(cConf, consumer.getName());
+            new DefaultMetadataConsumerContext(cConf, consumer.getName(), this.metricsCollectionService,
+                                               run.getNamespace());
+          MetadataConsumerMetrics metrics = metadataConsumerContext.getMetrics(Collections.emptyMap());
+          metrics.increment("metadata.consumer.calls.attempted", 1);
```
Comment (greeshmaswaminathan) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/metadata/MetadataConsumerSubscriberService.java:276]: Just to confirm , the intention is to capture attempted calls regardless of whether they are successful ? And no tags required  for the metric in this context?

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/metadata/MetadataConsumerSubscriberService.java [hunk 4]
```diff
@@ -265,35 +265,34 @@ public void processMessage(MetadataMessage message, StructuredTableContext conte
...
           // if there is any error from the implementation, log and continue here
           // as we are already retrying at the service level
           consumer.consumeLineage(metadataConsumerContext, run, info);
         } catch (Throwable e) {
-          LOG.error("Error calling the metadata consumer {}: {}", consumer.getName(),
-              e.getMessage());
+          LOG.error("Error calling the metadata consumer {}: {}", consumer.getName(), e.getMessage());
```
Comment (snehalakshmisha) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/metadata/MetadataConsumerSubscriberService.java:281]: Capture metrics for failed calls

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/metadata/DefaultMetadataConsumerContext.java [hunk 5]
```diff
@@ -16,32 +16,53 @@
...
   @Override
   public Map<String, String> getProperties() {
     return this.properties;
   }
+
+  @Override
+  public MetadataConsumerMetrics getMetrics(Map<String, String> context) {
+    Map<String, String> tags = new HashMap<>(context);
+    tags.put(Constants.Metrics.Tag.NAMESPACE, this.namespace);
+    tags.put(Constants.Metrics.Tag.METADATA_CONSUMER, metadataConsumerName);
```
Comment (snehalakshmisha) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/metadata/DefaultMetadataConsumerContext.java]: cehck if application name, program id and run id, etc are useful

Comment (snehalakshmisha) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/metadata/DefaultMetadataConsumerContext.java]: Useful to track success/failed calls for a pipeline.
Similar to - https://github.com/cdapio/cdap/blob/develop/cdap-app-fabric/src/main/java/io/cdap/cdap/internal/provision/DefaultProvisionerContext.java#L149

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/metadata/MetadataConsumerSubscriberService.java (hunk 3 above)
Comment (snehalakshmisha) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/metadata/MetadataConsumerSubscriberService.java:276]: Yes that's right

**Review by greeshmaswaminathan — APPROVED** [review]

**Pull request merged**